lib/rubocop/rspec/language/node_pattern.rb

```diff
@@ -6,29 +6,72 @@ module Language
       # Common node matchers used for matching against the rspec DSL
       module NodePattern
         extend RuboCop::NodePattern::Macros
+        extend RuboCop::RSpec::Language
+        include RuboCop::RSpec::Language::Config
 
         def_node_matcher :rspec?, '{(const {nil? cbase} :RSpec) nil?}'
 
-        def_node_matcher :example_group?, ExampleGroups::ALL.block_pattern
-        def_node_matcher :shared_group?, SharedGroups::ALL.block_pattern
+        def_node_matcher :example_group?,
+                         block_pattern('#rspec_all_example_groups')
 
-        spec_groups = ExampleGroups::ALL + SharedGroups::ALL
-        def_node_matcher :spec_group?, spec_groups.block_pattern
+        def_node_matcher :shared_group?,
+                         block_pattern('#rspec_all_shared_groups')
+
+        def_node_matcher :spec_group?,
+                         block_pattern(
+                           '{#rspec_all_shared_groups '\
+                           '#rspec_all_example_groups}'
+                         )
 
         def_node_matcher :example_group_with_body?, <<-PATTERN
-          (block #{ExampleGroups::ALL.send_pattern} args !nil?)
+          (block #{send_pattern('#rspec_all_example_groups')} args !nil?)
+        PATTERN
+
+        def_node_matcher :example?, block_pattern('#rspec_all_examples')
+
+        def_node_matcher :hook?, block_pattern('#rspec_hooks')
+
+        def_node_matcher :let?, <<-PATTERN
+          {
+            #{block_pattern('#rspec_helpers')}
+            #{block_pass_pattern('#rspec_helpers')}
+          }
+        PATTERN
+
+        def_node_matcher :include?, <<-PATTERN
+          {
+            #{send_pattern('#rspec_all_includes')}
+            #{block_pattern('#rspec_all_includes')}
+            #{block_pass_pattern('#rspec_all_includes')}
+          }
         PATTERN
 
-        def_node_matcher :example?, Examples::ALL.block_pattern
+        def_node_matcher :subject?, block_pattern('#rspec_subjects')
+
+        protected
 
-        def_node_matcher :hook?, Hooks::ALL.block_pattern
+        def rspec_language_for(*keys)
+          rspec_language_config.dig(*keys).to_a.map(&:to_sym)
+        end
 
-        def_node_matcher :let?, Helpers::ALL.block_or_block_pass_pattern
+        private
 
-        def_node_matcher :include?,
-                         Includes::ALL.send_or_block_or_block_pass_pattern
+        def rspec_all(keyword)
+          all_keywords.include?(keyword)
+        end
 
-        def_node_matcher :subject?, Subject::ALL.block_pattern
+        def all_keywords
+          @all_keywords ||= [
```

Comment:
Makes sense to move it to config 👍 

> When done, should be a Set
Could you elaborate?

Reply:
I guess @marcandre means that:
```ruby
@all_keywords ||= Set.new([
  all_example_groups_keywords,
  ...
```

This way it's possible to get rid of `.include?(keyword)` and the parameter here:
```ruby
        def rspec_all
          all_keywords
        end
```
There's one spec failing for `controller do`. Didn't dig further.

My point is to merge as is and then exercise our refactoring skills separately. We're in a good position for that with no 1.x releases left between the merge and 2.0.

Reply:
Hmm I was confused because `all_keywords` returns sum of `Set`s which is a `Set` too.

Reply:
Nice! 👍

Reply:
My mistake, sorry 😊

Reply:
The only way to memoize keyword `Set`s on config level that I currently see is to monkey-patch `RuboCop::Config`. I think no any facade or similar would work, config should have memoization on `RuboCop` level so that cops would be instantiated with the same instance of `Config`.
In this case it also makes sense to move there content of `RuboCop::Cop::RSpec::Base`.